ingestion/src/metadata/ingestion/source/objectstore/s3/metadata.py

```diff
@@ -0,0 +1,287 @@
+    Source as WorkflowSource,
+)
+from metadata.ingestion.api.source import InvalidSourceException
+from metadata.ingestion.source.database.datalake.metadata import DatalakeSource
+from metadata.ingestion.source.objectstore.objectstore_service import (
+    ObjectStoreServiceSource,
+)
+from metadata.utils.filters import filter_by_container
+from metadata.utils.logger import ingestion_logger
+from metadata.utils.s3_utils import prefix_exits
+
+logger = ingestion_logger()
+
+OPENMETADATA_TEMPLATE_FILE_NAME = "openmetadata.json"
+
+
+class S3Metric(Enum):
+    NUMBER_OF_OBJECTS = "NumberOfObjects"
+    BUCKET_SIZE_BYTES = "BucketSizeBytes"
+
+
+class S3BucketResponse(BaseModel):
+    """
+    Class modelling a response received from s3_client.list_buckets operation
+    """
+
+    class Config:
+        extra = Extra.forbid
+
+    Name: str = Field(..., description="Bucket name", title="Bucket Name")
+    CreationDate: datetime = Field(
+        ...,
+        description="Timestamp of Bucket creation in ISO format",
+        title="Creation Timestamp",
+    )
+
+
+class S3ContainerDetails(BaseModel):
+    class Config:
+        extra = Extra.forbid
+
+    name: str = Field(..., description="Bucket name", title="Bucket Name")
+    numberOfObjects: float = Field(..., description="Total nr. of objects", title="Nr. of objects")
+    size: float = Field(..., description="Total size in bytes of all objects", title="Total size(bytes) of objects")
+    fileFormats: Optional[List[container.FileFormat]] = Field(..., description="File formats", title="File formats")
+    dataModel: Optional[ContainerDataModel] = Field(..., description="Data Model for the container", title="Data Model")
+    CreationDate: str = Field(
+        ...,
+        description="Timestamp of Bucket creation in ISO format",
+        title="Creation Timestamp",
+    )
+
+
+class S3Source(ObjectStoreServiceSource):
+    """
+    Source implementation to ingest S3 buckets data.
+    """
+
+    def __init__(self, config: WorkflowSource, metadata_config: OpenMetadataConnection):
+        super().__init__(config, metadata_config)
+        self.s3_client = self.connection.s3_client
+        self.cloudwatch_client = self.connection.cloudwatch_client
+
+    @classmethod
+    def create(cls, config_dict, metadata_config: OpenMetadataConnection):
+        config: WorkflowSource = WorkflowSource.parse_obj(config_dict)
+        connection: S3StoreConnection = config.serviceConnection.__root__.config
+        if not isinstance(connection, S3StoreConnection):
+            raise InvalidSourceException(
+                f"Expected S3StoreConnection, but got {connection}"
+            )
+        return cls(config, metadata_config)
+
+    def get_containers(self) -> Iterable[S3ContainerDetails]:
+        """
+        List and filter containers
+        """
+        buckets = []
+        try:
+            buckets = self.s3_client.list_buckets()['Buckets']
```

Comment:
Should we `.get("Buckets")` and handle `None` in buckets?

Reply:
rather than allowing this to blow up in a KeyError

Reply:
Used get and extra check for None before iterating the list of buckets.